rust-src/concordium_base/src/id/sigma_protocols/com_ineq.rs

```diff
@@ -0,0 +1,163 @@
+    let mut diff = pub_value;
+    diff.negate();
+    diff.add_assign(value);
+
+    // compute the inverse of (value - pub_value)
+    let diff_inv = diff.inverse()?;
+
+    // generate commitment cmm_1 to `diff` using the given randomness `value_tilde`;
+    // this allows the verifier to compute cmm_1 from public commitment
+    // c=g^{value} h^{value_tilde} for com_key=(g, h).
+    let diff_val = Value::new(diff);
+    let diff_inv_val = Value::new(diff_inv);
+    let cmm_1 = com_key.hide(&diff_val, &value_tilde);
+    let (cmm_2, r_2) = com_key.commit(&diff_inv_val, csprng);
+    //  Compute a commitment to 1. Alternatively, one could use cmm_3 =
+    // com_key.hide(one, zero).
+    let cmm_3 = Commitment(com_key.g);
+
+    // This is the com-mult proof in section 9.2.10 from the Bluepaper.
+    let prover = ComMult {
+        cmms:    [cmm_1, cmm_2, cmm_3],
+        cmm_key: *com_key,
+    };
+
+    let secret = ComMultSecret {
+        values: [diff_val, diff_inv_val],
+        rands:  [value_tilde, r_2, Randomness::<C>::zero()],
+    };
+
+    let partial_proof = sigma_prove(&mut transcript, &prover, secret, csprng);
+    let partial_proof = partial_proof.unwrap();
+    Some(Witness {
+        com_mult_witness: partial_proof,
+        aux_com:          cmm_2,
+    })
+}
+
+pub fn verify_com_ineq<C: Curve>(
+    com_key: &CommitmentKey<C>,
+    c: &Commitment<C>,
+    pub_value: C::Scalar,
+    proof: &Witness<C>,
+) -> bool {
+    let Witness {
+        com_mult_witness,
+        aux_com,
+    } = proof;
+    let mut transcript = RandomOracle::domain(b"NonEqualityProof");
+    transcript.append_message(b"g", &com_key.g);
+    transcript.append_message(b"h", &com_key.h);
+    transcript.append_message(b"public commitment", &c);
+    transcript.append_message(b"public value", &pub_value);
+    let mut minus_pub_value = pub_value;
+    minus_pub_value.negate();
+    let g_minus_pub_value = com_key.g.mul_by_scalar(&minus_pub_value);
+
+    let cmm_1: C = c.plus_point(&g_minus_pub_value);
+    let cmm_3 = Commitment(com_key.g);
+
+    let com_mult = ComMult {
+        cmms:    [Commitment(cmm_1), *aux_com, cmm_3],
+        cmm_key: *com_key,
+    };
+    sigma_verify(&mut transcript, &com_mult, com_mult_witness)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ff::PrimeField;
+    use pairing::bls12_381::{Fr, G1};
+
+    #[test]
+    fn test_com_ineq_correctness() {
+        let mut csprng = rand::thread_rng();
+        let com_key: CommitmentKey<_> = CommitmentKey::<G1>::generate(&mut csprng);
+
+        let value = Value::<G1>::new(Fr::from_str("20000102").unwrap());
+        let pub_value = Fr::from_str("20000103").unwrap();
+
+        let (cmm_1, value_tilde) = com_key.commit(&value, &mut csprng);
+
+        let proof = prove_com_ineq(&com_key, &value, value_tilde, pub_value, &mut csprng)
+            .expect("Proving should succeed.");
+        assert!(
+            verify_com_ineq(&com_key, &cmm_1, pub_value, &proof),
```

Comment:
Why does this succeed if different commitment is used then in proof? Shouldn't this fail because the statement is about another commitment?